features/frameworks/mbed-events/EventQueue.h

```diff
@@ -0,0 +1,764 @@
+/* events
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef EVENT_QUEUE_H
+#define EVENT_QUEUE_H
+
+#include "equeue/equeue.h"
+#include "Callback.h"
+#include <cstddef>
+#include <new>
+
+namespace events {
+
+/** EVENTS_EVENT_SIZE
+ *  Minimum size of an event
+ *  This size fits a Callback<void()> at minimum
+ */
+#define EVENTS_EVENT_SIZE \
+    (EQUEUE_EVENT_SIZE - 2*sizeof(void*) + sizeof(mbed::Callback<void()>))
+
+/** EVENTS_QUEUE_SIZE
+ *  Default size of buffer for events
+ */
+#define EVENTS_QUEUE_SIZE (32*EVENTS_EVENT_SIZE)
+
+// Predeclared classes
+template <typename F>
+class Event;
+
+
+/** EventQueue
+ *
+ *  Flexible event queue for dispatching events
+ */
+class EventQueue {
+public:
+    /** EventQueue lifetime
+     *
+     *  Create and destroy event queues. The event queue either allocates
+     *  a buffer of the specified size with malloc or uses the user provided
+     *  buffer.
+     *
+     *  @param size     Size of buffer to use for events in bytes
+     *                  (default to EVENTS_QUEUE_SIZE)
+     *  @param buffer   Pointer to buffer to use for events
+     *                  (default to NULL)
+     */
+    EventQueue(unsigned size=EVENTS_QUEUE_SIZE, unsigned char *buffer=NULL);
+    ~EventQueue();
+
+    /** Dispatch events
+     *
+     *  Executes events until the specified milliseconds have passed.
+     *  If ms is negative, the dispatch function will dispatch events
+     *  indefinitely or until break_dispatch is called on this queue.
+     *
+     *  When called with a finite timeout, the dispatch function is garunteed
+     *  to terminate. When called with a timeout of 0, the dispatch function
+     *  does not wait and is irq safe.
+     *
+     *  @param ms       Time to wait for events in milliseconds, a negative
+     *                  value will dispatch events indefinitely
+     *                  (default to -1)
+     */
+    void dispatch(int ms);
+    void dispatch_forever() { dispatch(-1); }
```

Comment:
Looked into it again, looks like you're right, the inference in the `callback` function is causing an ambiguity with the overloads. Looks like overloaded functions may have some issues with the callback class.

You can force the overload resolution to the expected callback type explicitly as a general solution:
``` cpp
t.start(Callback<void()>(&e, &EventQueue::dispatch));
```

Or your `dispatch_forever` member function provides an ok workaround:
``` cpp
t.start(callback(&e, &EventQueue::dispatch_forever));
```

I'll let you decide what to go with, I put a pr up for `dispatch_forever` to keep these in synch if you think this is cleaner (https://github.com/ARMmbed/mbed-events/pull/20) 

It would be nice to keep the `dispatch(int ms=-1)` default argument.